File: backend/app/schemas/blog.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime


# Blog Folder Schemas
class BlogFolderBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    icon: str = Field(default="folder", max_length=50)
    order_index: int = Field(default=0)

class BlogFolderCreate(BlogFolderBase):
    parent_id: Optional[str] = None

class BlogFolderUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    icon: Optional[str] = Field(None, max_length=50)
    order_index: Optional[int] = None
    parent_id: Optional[str] = None
    is_expanded: Optional[bool] = None

class BlogFolderResponse(BlogFolderBase):
    id: str
    user_id: str
    parent_id: Optional[str]
    is_expanded: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Blog Post Schemas
class BlogPostBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    slug: str = Field(..., min_length=1, max_length=255)
    content_md: str
    content_blocks: Optional[List[Any]] = None
    excerpt: Optional[str] = None
    cover_image: Optional[str] = None
    tags: Optional[List[str]] = Field(default_factory=list)
    status: str = Field(default="draft", max_length=20)

class BlogPostCreate(BlogPostBase):
    series_id: Optional[str] = None
    folder_id: Optional[str] = None
    github_repo_id: Optional[str] = None

class BlogPostUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    slug: Optional[str] = Field(None, min_length=1, max_length=255)
    content_md: Optional[str] = None
    content_blocks: Optional[List[Any]] = None
    excerpt: Optional[str] = None
    cover_image: Optional[str] = None
    tags: Optional[List[str]] = None
    status: Optional[str] = Field(None, max_length=20)
    series_id: Optional[str] = None
    folder_id: Optional[str] = None
    github_repo_id: Optional[str] = None

class BlogPostResponse(BlogPostBase):
    id: str
    user_id: str
    series_id: Optional[str]
    folder_id: Optional[str]
    github_repo_id: Optional[str]
    github_path: Optional[str]
    github_sha: Optional[str]
    published_at: Optional[datetime]
    view_count: int
    likes_count: int = 0
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Public schemas for feed/public API
class AuthorPublic(BaseModel):
    id: str
    username: str
    github_username: Optional[str] = None
    avatar_url: Optional[str] = None
    bio: Optional[str] = None

    class Config:
        from_attributes = True


class BlogPostPublic(BaseModel):
    id: str
    title: str
    slug: str
    excerpt: Optional[str] = None
    cover_image: Optional[str] = None
    tags: Optional[List[str]] = None
    status: str
    published_at: Optional[datetime]
    view_count: int
    likes_count: int = 0
    comments_count: int = 0
    created_at: datetime
    author: AuthorPublic

    class Config:
        from_attributes = True


class BlogPostDetailPublic(BlogPostPublic):
    content_md: str
    content_blocks: Optional[List[Any]] = None
    github_repo_id: Optional[str] = None

    class Config:
        from_attributes = True

class SeriesBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    slug: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    thumbnail: Optional[str] = None
    is_public: bool = Field(default=True)

class SeriesCreate(SeriesBase):
    pass

class SeriesUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    slug: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    thumbnail: Optional[str] = None
    is_public: Optional[bool] = None

class SeriesResponse(SeriesBase):
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Comment Schemas
class CommentAuthor(BaseModel):
    id: str
    username: str
    github_username: Optional[str] = None
    avatar_url: Optional[str] = None

    class Config:
        from_attributes = True


class CommentCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=5000)
    parent_id: Optional[str] = None


class CommentUpdate(BaseModel):
    content: str = Field(..., min_length=1, max_length=5000)


class CommentResponse(BaseModel):
    id: str
    content: str
    author: CommentAuthor
    parent_id: Optional[str] = None
    is_deleted: bool
    created_at: datetime
    updated_at: datetime
    replies: List["CommentResponse"] = []

    class Config:
        from_attributes = True


# Stats History for charts
class DailyStats(BaseModel):
    date: str
    views: int = 0
    likes: int = 0
    comments: int = 0


class StatsHistory(BaseModel):
    daily_stats: List[DailyStats]
    total_views: int
    total_likes: int
    total_comments: int

File: frontend/src/types/blog.ts
export interface BlogFolder {
  id: string;
  user_id: string;
  parent_id: string | null;
  name: string;
  icon: string;
  order_index: number;
  is_expanded: boolean;
  created_at: string;
  updated_at: string;
  children?: BlogFolder[];
  posts?: BlogPost[];
}

export interface BlogPost {
  id: string;
  user_id: string;
  series_id: string | null;
  folder_id: string | null;
  title: string;
  slug: string;
  content_md: string;
  content_blocks: any[] | null;
  excerpt: string | null;
  cover_image: string | null;
  tags: string[] | null;
  status: 'draft' | 'published' | 'archived';
  published_at: string | null;
  view_count: number;
  likes_count: number;
  github_repo_id: string | null;
  github_path: string | null;
  github_sha: string | null;
  created_at: string;
  updated_at: string;
}

// Public API types
export interface AuthorPublic {
  id: string;
  username: string;
  github_username: string | null;
  avatar_url: string | null;
  bio: string | null;
}

export interface BlogPostPublic {
  id: string;
  title: string;
  slug: string;
  excerpt: string | null;
  cover_image: string | null;
  tags: string[] | null;
  status: string;
  published_at: string | null;
  view_count: number;
  likes_count: number;
  comments_count: number;
  created_at: string;
  author: AuthorPublic;
}

export interface BlogPostDetailPublic extends BlogPostPublic {
  content_md: string;
  content_blocks: any[] | null;
  github_repo_id: string | null;
}

// Social link type (moved up for use in UserPublicProfile)
export interface SocialLink {
  platform: string;
  url: string;
  label?: string;
}

export interface UserPublicProfile {
  id: string;
  username: string;
  github_username: string | null;
  avatar_url: string | null;
  bio: string | null;
  social_links: SocialLink[] | null;
  is_github_connected: boolean;
  created_at: string;
  followers_count: number;
  following_count: number;
  posts_count: number;
  is_following: boolean;
}

export interface BlogSeries {
  id: string;
  user_id: string;
  title: string;
  slug: string;
  description: string | null;
  thumbnail: string | null;
  is_public: boolean;
  created_at: string;
  updated_at: string;
}

export interface GitHubRepo {
  id: string;
  github_repo_id: string;
  name: string;
  full_name: string;
  description: string | null;
  url: string;
  language: string | null;
  stars_count: number;
  forks_count: number;
  is_private: boolean;
  topics: string[] | null;
}

export interface FolderTreeItem {
  id: string;
  name: string;
  type: 'folder' | 'post';
  icon: string;
  parent_id: string | null;
  is_expanded?: boolean;
  children?: FolderTreeItem[];
  post?: BlogPost;
}

// My Page types
export interface UserStats {
  total_posts: number;
  total_views: number;
  total_likes_received: number;
  total_likes_given: number;
  followers_count: number;
  following_count: number;
}

export interface MyPageProfile {
  id: string;
  username: string;
  email: string;
  github_username: string | null;
  avatar_url: string | null;
  bio: string | null;
  social_links: SocialLink[] | null;
  is_github_connected: boolean;
  created_at: string;
  stats: UserStats;
}

export interface FollowerUser {
  id: string;
  username: string;
  github_username: string | null;
  avatar_url: string | null;
  bio: string | null;
  is_following: boolean;
}

// Comment types
export interface CommentAuthor {
  id: string;
  username: string;
  github_username: string | null;
  avatar_url: string | null;
}

export interface Comment {
  id: string;
  content: string;
  author: CommentAuthor;
  parent_id: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  replies: Comment[];
}

// Stats History types
export interface DailyStats {
  date: string;
  views: number;
  likes: number;
  comments: number;
}

export interface StatsHistory {
  daily_stats: DailyStats[];
  total_views: number;
  total_likes: number;
  total_comments: number;
}

// Notification types
export interface NotificationActor {
  id: string;
  username: string;
  github_username: string | null;
  avatar_url: string | null;
  is_following: boolean;
}

export interface NotificationPost {
  id: string;
  title: string;
  slug: string;
  author_username: string;
}

export interface NotificationComment {
  id: string;
  content: string;
  post_id: string;
}

export interface Notification {
  id: string;
  type: 'like' | 'comment' | 'follow' | 'reply';
  actor: NotificationActor;
  post: NotificationPost | null;
  comment: NotificationComment | null;
  is_read: boolean;
  created_at: string;
}

export interface NotificationList {
  notifications: Notification[];
  unread_count: number;
  total_count: number;
}
